src/CMakeLists.txt

```diff
@@ -0,0 +1,149 @@
+# Must be included before CMAKE_INSTALL_INCLUDEDIR is used.
+include(GNUInstallDirs)
+set(${PROJECT_NAME}_installables "")
+
+if(SECP256K1_ASM STREQUAL "arm")
+  add_library(common OBJECT
+    asm/field_10x26_arm.s
+  )
+  set(common_obj "$<TARGET_OBJECTS:common>")
+else()
+  set(common_obj "")
+endif()
+
+add_library(precomputed OBJECT
+  precomputed_ecmult.c
+  precomputed_ecmult_gen.c
+)
+set(internal_obj "$<TARGET_OBJECTS:precomputed>" "${common_obj}")
+
+add_library(${PROJECT_NAME} SHARED EXCLUDE_FROM_ALL
```

Comment:
This static/shared logic seems needlessly complicated to me, and from what I can tell largely bypasses CMake's typical machinery and customs.

Is the extra work here primarily to support building a shared/static lib at the same time? If so, is that something we really need to carry forward?

Reply:
> Is the extra work here primarily to support building a shared/static lib at the same time? If so, is that something we really need to carry forward?

I agree. If this is only done to support both types at the same time, this is something we could drop for sure. No need to mimic that specific ability of autotools. If one needs both, one can invoke the build system twice.

Reply:
The PIC issue above is that the object lib doesn't know whether or not it should be built as PIC, so it defaults to no.
For static libs, that's the correct answer. For shared libs, that's wrong.

But since it's only built once no matter what, there's no correct answer to "how should the object lib be built if we're building for both shared and static libs?". The result is the hacky patch above.

So while I appreciate the attempt to do both at once, I think the only real "clean" way is to build objects twice. Which as @real-or-random said, means simply invoking twice.

In case it's helpful, yesterday I pushed a draft [CMake port of Minisketch](https://github.com/theuni/minisketch/commit/06a51b739da9faa8cb9f7f5c4765be090b657d7f) which I believe handles this correctly (though it currently uses `BUILD_SHARED_LIBS` directly). Notice how it gets/sets the `POSITION_INDEPENDENT_CODE` properly for the object lib so that it does the right thing whether it's building shared or static.

Reply:
> This static/shared logic seems needlessly complicated to me, and from what I can tell largely bypasses CMake's typical machinery and customs.
> 
> Is the extra work here primarily to support building a shared/static lib at the same time?

Yes, it is.



> The PIC issue above is that the object lib doesn't know whether or not it should be built as PIC, so it defaults to no.
> For static libs, that's the correct answer. For shared libs, that's wrong.

That is why there was `set(CMAKE_POSITION_INDEPENDENT_CODE ON)` [initially](https://github.com/bitcoin-core/secp256k1/pull/1113#discussion_r1124264718).

Reply:
> In case it's helpful, yesterday I pushed a draft [CMake port of Minisketch](https://github.com/theuni/minisketch/commit/06a51b739da9faa8cb9f7f5c4765be090b657d7f)...

FWIW, https://github.com/sipa/minisketch/pull/75

Reply:
> So while I appreciate the attempt to do both at once, I think the only real "clean" way is to build objects twice. Which as @real-or-random said, means simply invoking twice.

Agree.

FWIW, [here](https://discourse.cmake.org/t/one-library-target-building-both-static-and-shared/3155) are some related thoughts from one of CMake developers.

Reply:
> That is why there was `set(CMAKE_POSITION_INDEPENDENT_CODE ON)` [initially](https://github.com/bitcoin-core/secp256k1/pull/1113#discussion_r1124264718).

Makes sense! Sorry, I wasn't aware of all these issues when I suggested dropping that line.

I think for now, we should either
1. apply the patch suggested by @theuni 
2. or simply `set(CMAKE_POSITION_INDEPENDENT_CODE ON)` again -- sure it's a deviation from autotools then, but it's justified

If you ask me, I'd probably go with 1 but either is fine.

It's totally reasonable to say that dropping the ability to build shared and static libs at the same time belongs to the category of making things more CMake-ish and is probably better addressed in a separate PR. 

The big advantage of getting this PR merged soon that it will be easier to refine separate aspects of the CMake build separately. This PR here is huge already now (with 240 comments and 47 revisions), and discussions will get much easier if they can happen in separate PRs which address only a single concern.

Reply:
Oh. you already implemented suggestion 1 while I was writing this. :)